### DISCUSSION TITLE
Acquisition values always ZERO

### DISCUSSION THREAD

**Discussion by renjieli08** [meta-pytorch/botorch#1846] [Q&A] [answered]
Hi,
I've been using botorch to optimize an 8-input and 5-output objective function, essentially following the code in this tutorial: [https://botorch.org/tutorials/constrained_multi_objective_bo](url). The parallel Noisy Expected Hypervolume Improvement (𝑞NEHVI) is chosen as the acquisition function as in the tutorial. The only difference is that my optimization problem doesn't have a constraint. 

**Now my issue is that the acquisition function values are constantly zero**, not matter how I adjust my model and hyper-parameters. Here is my code: 

```
def generate_initial_data():
      # generate training data
      #input variable: state
    train_X = torch.zeros(5, 8, device=device, dtype=torch.float64)  #fixed zero initialization
      
    train_Y = torch.zeros(5,5, device=device, dtype=torch.float64)
    for i in range(0, 5):
        obj, _ = env.step(train_X[i, :].tolist())  #objective function
        # output variable: score
        train_Y[i, :] = torch.tensor(obj, device=device)

    return train_X, train_Y

def initialize_model(train_X, train_Y):
    # define models for objective and constraint

    d = 8  #No. of input parameters
    m = 5  #No. of objectives
    theta = 3   #lengthscale
    #define the matern kernel with custom lengthscale. Default lengthscale = 1.
    covariance = ScaleKernel(
        base_kernel=MaternKernel(
        nu=2.5,
        ard_num_dims=d,
        lengthscale_prior=GammaPrior(6.0, 9.0),
        )
    )

    #use model list for improved performance
    models = []
    for i in range(m):
        models.append(
            SingleTaskGP(train_X, train_Y[..., i : i + 1], covar_module= covariance,
                      outcome_transform=Standardize(m=1), input_transform=Normalize(d=d)
            )
        )
    model = ModelListGP(*models)
    mll = SumMarginalLogLikelihood(model.likelihood, model)
    
    return mll, model

def optimize_acqf_and_get_observation(model, train_x):
    """Optimizes the acquisition function, and returns a new candidate and a noisy observation."""

    #bounds for the x input
    high = [1000.0, 200., 200., 200., 0.15, 0.15, 0.3, 100.]
    low = [-1000.0, -200., -200., -200., -0.15, -0.15, -0.3, -100.]
    bounds = torch.tensor([low, high], dtype=torch.float64)

    train_x = normalize(train_x, bounds)

    #acquisition function (qNEHVI for multi-objective)
    acq_func = qNEHVI(
            model=model,
            ref_point= torch.zeros(5),   #lower bound for objectives
            X_baseline=train_x,
            prune_baseline=True,
            alpha=1.0,
    )

    print('\nacq func')
    print(acq_func(train_x))

    #optimize
    candidate, acq_value = optimize_acqf(
        acq_function = acq_func, bounds=bounds, q=1, num_restarts=5, raw_samples=20,
    )

    # observe new values
    new_x = candidate.detach()
    #new_x = torch.round(new_x)
    obj, score = env.step(new_x.tolist()[0])  #objective function 
    new_y = torch.tensor(obj, device=device).unsqueeze(0)
    return new_x, new_y, acq_value, score


N_TRIALS = 60
N_BATCH = 300

verbose = True

#save overall best solutions
best_observed_all_ei = []

#save data
input = torch.tensor([])
objective = torch.tensor([])

steps_done = 0

#average over multiple trials
for trial in range(1, N_TRIALS + 1):
    utc_dt = datetime.now(timezone.utc)

    print(f"\nTrial {trial:>2} of {N_TRIALS} \n", end="")

    print("\nLocal time {}".format(utc_dt.astimezone().isoformat()))
    
    best_observed_ei = []

    #define the hypervolume
    hv = Hypervolume(ref_point=torch.zeros(5))
    hvs_qnehvi = []

    # call helper functions to generate initial training data and initialize model
    (
        train_x_ei,
        train_obj_ei,
    ) = generate_initial_data()

    #print('\nx_initial: {}, f_initial: {:.5f}\n'.format(train_x_ei.tolist()[0], train_obj_ei.tolist()[0]))
    print('f_initial:')
    print(train_obj_ei.tolist())
    print('\nx_initial:')
    print(train_x_ei.tolist())

    mll_ei, model_ei = initialize_model(train_x_ei, train_obj_ei)

    best_observed_ei.append(train_obj_ei)

    # compute pareto front
    feas_train_obj = train_obj_ei
    if feas_train_obj.shape[0] > 0:
        pareto_mask = is_non_dominated(feas_train_obj)
        pareto_y = feas_train_obj[pareto_mask]
        # compute hypervolume
        volume = hv.compute(pareto_y)
    else:
        volume = 0.0

    hvs_qnehvi.append(volume)

    # run N_BATCH rounds of BayesOpt after the initial random batch
    for iteration in range(1, N_BATCH + 1):
        print('\nStarting step No.{}'.format(iteration))

        steps_done += 1

        # fit the models
        fit_gpytorch_mll(mll_ei)

        # print('\n lengthscale')
        # print(model_ei.covar_module.base_kernel.lengthscale)

        # optimize and get new observation
        new_x_ei, new_obj_ei, acq, score = optimize_acqf_and_get_observation(model_ei, train_x_ei)

        print('f_obj, f_score:')
        print(new_obj_ei.tolist(), score)
        print('\nx_candidate:')
        print(new_x_ei.tolist())
        print('\nacquisition:')
        print(acq.tolist())

        posterior = model_ei.posterior(new_x_ei)
        mean = posterior.mean  #posterior mean
        # Get upper and lower confidence bounds (2 std dev from the mean)
        lower, upper = posterior.mvn.confidence_region()

        print('\nPosterior mean and uncertainty:')
        print(mean.tolist(), upper.tolist())

        #writer.add_scalar('BO/scores', new_obj_ei.item(), steps_done)
        writer.add_scalar('BO/acq value', acq.item(), steps_done)
        writer.add_scalar('BO/f score', score, steps_done)

        # update training points
        train_x_ei = torch.cat([train_x_ei, new_x_ei])
        train_obj_ei = torch.cat([train_obj_ei, new_obj_ei])

        # update progress
        for hvs_list, train_obj in zip(
            (hvs_qnehvi,),
            (train_obj_ei,),
        ):
            # compute pareto front
            feas_train_obj = train_obj.unsqueeze(0)
            if feas_train_obj.shape[0] > 0:
                pareto_mask = is_non_dominated(feas_train_obj)
                pareto_y = feas_train_obj[pareto_mask]
                # compute feasible hypervolume
                volume = hv.compute(pareto_y)
            else:
                volume = 0.0
            hvs_list.append(volume)

        print('\nhypervolume:')
        print(volume)
        writer.add_scalar('BO/hypervolume', volume, steps_done)

        # reinitialize the models so they are ready for fitting on next iteration
        # use the current state dict to speed up fitting
        mll_ei, model_ei = initialize_model(
            train_x_ei,
            train_obj_ei,
        )


        if iteration % 50 == 0:
            #save data
            input = torch.cat([input, train_x_ei])
            objective = torch.cat([objective, train_obj_ei])
            torch.save(input, 'input.pt')
            torch.save(objective, 'objective.pt')

        if verbose:
            print(
                f"Batch {iteration:>2}: Hypervolume (qNEHVI) = "
                f"({hvs_qnehvi[-1]:>4.5f}), ",
                end="\n",
            )
        else:
            print(".", end="\n")
```

**Here is a sample output from running the above code:**
BadInitialCandidatesWarning: Unable to find non-zero acquisition function values - initial conditions are being selected randomly.

Starting step No.1

acq func
tensor([0.], dtype=torch.float64, grad_fn=<AddBackward0>)

Q factor: 14740.106, resonance lambda: 1406.38, area: 8.5639e-14, power: 0.0000, divergence: 1.1614

f_obj, f_score:
[[0.0073700529412966675, 0.9264244075447232, 0.23788713235941283, 5.672858516270196e-09, 0.8385756450397841]] 41.49402116450884

x_candidate:
[[-79.38002981245518, 32.086195796728134, -148.4838206321001, -35.77237389981747, 0.10972788669168951, 0.05081366663798689, -0.05658666919916869, -52.36530061811209]]

acquisition:
0.0

Posterior mean and uncertainty:
[[0.05059579170492109, 0.9492523232296569, 0.2618673535516844, 0.3117851135315177, 0.838471422626162]] [[0.050653100885836064, 0.9492723232296569, 0.2618873535516844, 0.3119949052921522, 0.838491422626162]]

hypervolume:
0.017218729481101036
Batch  1: Hypervolume (qNEHVI) = (0.01722), 

Starting step No.2
BadInitialCandidatesWarning: Unable to find non-zero acquisition function values - initial conditions are being selected randomly.

acq func
tensor([0.], dtype=torch.float64, grad_fn=<AddBackward0>)

Q factor: 133590.653, resonance lambda: 1362.14, area: 8.4514e-14, power: 0.0003, divergence: 1.1615

f_obj, f_score:
[[0.06679532659907061, 0.9601957026595896, 0.2347601551320646, 0.00095828431855427, 0.8384761202505044]] 41.36835737933064

x_candidate:
[[-4.69064898788929, -147.21730574965477, -81.92704394459724, -79.19720783829689, -0.07687873654067516, 0.044364464934915304, 0.10664636492729185, -97.85456620156765]]

acquisition:
0.0

Posterior mean and uncertainty:
[[0.04339396979224573, 0.94544767050204, 0.2578706497836072, 0.25982092815776137, 0.8384887930284166]] [[0.04341396979224573, 0.9454676705020401, 0.2578906497836072, 0.2598667985236072, 0.8385087930284166]]

hypervolume:
0.01721883937716484
Batch  2: Hypervolume (qNEHVI) = (0.01722), 

**As you can see above, the acquisition values are zero**, and I get the BadInitialCandidatesWarning warning each time.
On a side note, the hypervolumes and posterior means seem correct.
Any idea what could have gone wrong? Help is really appreciated !!

**Comment by Balandat** [comment, marked as answer]
Hmm so my first guess here is that this is b/c of the choice of `alpha=1.0` - this is a parameter controlling the coarseness of the hypervolume computation, and typically chosen rather small since it expresses a fraction. From https://github.com/pytorch/botorch/blob/main/botorch/utils/multi_objective/box_decompositions/non_dominated.py#L48-L51 setting it to 1.0 suggests the HV will always be trivially zero. Try setting it to a small value, such as 0.05 or 0.1. 

But note that with 5 outcomes the complexity of the HV computation can be quite high, so this could take a while. At that point if you really do care about the tradeoffs between all outcomes you could consider some other scalarization-based methods. But I'm curious how you're expecting to process / use the Pareto frontier in 5 dimensions, and whether this is indeed a full multi-objective problem or whether a MOO formulation with fewer objectives but some constraints would make sense.

Finally, we have some work that improves the numerical behavior of the improvement-based acquisition functions, we hope to open source that soon. This seems to be a problem that could also benefit quite a bit from this. 

cc @sdaulton, @SebastianAment 

 


 I don't see anything obviously wrong with this, assuming that the reference point at zero here is reasonably conservative / that all outcomes are non-negative as the comment suggests.

**Reply by renjieli08** [reply]
Hi there,
Appreciate your timely reply. Here is the output after I changed alpha from 1 to 0:

Starting step No.10

acq func
C:\Users\haihuang\Miniconda3\envs\fdtd\lib\site-packages\linear_operator\utils\cholesky.py:40: NumericalWarning: A not p.d., added jitter of 1.0e-08 to the diagonal
  warnings.warn(
tensor([7.4032685780e-06], dtype=torch.float64, grad_fn=<AddBackward0>)
starting new FDTD session...

Q factor: -21585.623, resonance lambda: 1327.53, area: 8.7055e-14, power: 0.0014, divergence: 1.1615

f_obj, f_score:
[[-0.010792811724270823, 0.986620588725405, 0.24182011735822706, 0.004585623923119342, 0.8385247021414537]] 42.386895947964845

x_candidate:
[[931.7723251879215, 117.38879606127739, 24.297846481204033, 14.8428525775671, -0.07457528114318847, -0.1238873055204749, 0.29943717867136005, -48.539046943187714]]

acquisition:
2.1387586746517376e-07
C:\Users\haihuang\Miniconda3\envs\fdtd\lib\site-packages\gpytorch\distributions\multivariate_normal.py:319: NumericalWarning: Negative variance values detected. This is likely
due to numerical instabilities. Rounding negative variances up to 1e-10.
  warnings.warn(

Posterior mean and uncertainty:
[[0.1135903302831784, 0.9532514332899666, 0.2368104948095983, 0.16570317373789678, 0.838484256372174]] [[0.11383293991903519, 0.9532714332899667, 0.23684833479370568, 0.16585578187913383, 0.8385042563721741]]

hypervolume:
0.019333185628056526
Batch 10: Hypervolume (qNEHVI) = (0.01933),

Starting step No.11

acq func
C:\Users\haihuang\Miniconda3\envs\fdtd\lib\site-packages\linear_operator\utils\cholesky.py:40: NumericalWarning: A not p.d., added jitter of 1.0e-08 to the diagonal
  warnings.warn(
tensor([6.3484794429e-06], dtype=torch.float64, grad_fn=<AddBackward0>)
starting new FDTD session...

Q factor: 1655471.972, resonance lambda: 1348.44, area: 8.9185e-14, power: 0.0098, divergence: 1.1615

f_obj, f_score:
[[0.8277359858800497, 0.9706574848168468, 0.24773579372723298, 0.032501778725411934, 0.8385261951941789]] 46.59267461984497

x_candidate:
[[-753.9003789424896, 137.26902455091476, 24.798031523823738, 155.31849153339863, -0.12768304729834198, -0.05982640692964196, 0.26400913037359713, 31.833583675324917]]

acquisition:
2.1065067677286072e-07
C:\Users\haihuang\Miniconda3\envs\fdtd\lib\site-packages\gpytorch\distributions\multivariate_normal.py:319: NumericalWarning: Negative variance values detected. This is likely
due to numerical instabilities. Rounding negative variances up to 1e-10.
  warnings.warn(

Posterior mean and uncertainty:
[[0.1053008990645664, 0.9554760433997954, 0.23714447045978354, 0.15496200862309312, 0.8384869527564519]] [[0.10556030090144049, 0.9554960433997954, 0.23718471349038067, 0.1551281692321272, 0.8385069527564519]]

hypervolume:
0.022425130009651184
Batch 11: Hypervolume (qNEHVI) = (0.02243),

**As you can see above, acquisition values are indeed no longer zero, albeit rather small (e.g., on the order of ~ 2.0e-07). Are acquisition values that small something we would normally expect?** 

Besides, I noticed that my posterior mean and 2 standard deviation confidence are almost identical, as seen in the output. I don't think this is a coincidence, and if not, what could have been wrong here?

In addition, in the output you can see two types of **numerical warnings:** a "not p.d., added jitter" warning and a "Negative variance values" warning. Any idea how to prevent those from occurring?

Also, 
> But I'm curious how you're expecting to process / use the Pareto frontier in 5 dimensions, and whether this is indeed a full multi-objective problem or whether a MOO formulation with fewer objectives but some constraints would make sense.

In the quote above, are you suggesting that 5 objectives is too much for Botorch to handle? I am asking cuz I've read in the literature that MOBO can generally handle upto 15-ish objectives. 

Also, I don't believe my problem has a constraint, other than a simple box bound imposed on the input parameters that has been explicitly written out in the optimize_acqf_and_get_observation(model, train_x) method in my code :
```
high = [1000.0, 200., 200., 200., 0.15, 0.15, 0.3, 100.]
low = [-1000.0, -200., -200., -200., -0.15, -0.15, -0.3, -100.]
bounds = torch.tensor([low, high], dtype=torch.float64)
```

Thanks again.

**Comment by Balandat** [comment]
> As you can see above, acquisition values are indeed no longer zero, albeit rather small (e.g., on the order of ~ 2.0e-07). Are acquisition values that small something we would normally expect?

Yes it's possible - you're integrating an area in 5 dimensions w.r.t. to a MVN with light tails, this can produce rather small numbers. The versions that @SebastianAment has implemented should be helpful here. 

> In addition, in the output you can see two types of numerical warnings: a "not p.d., added jitter" warning and a "Negative variance values" warning. Any idea how to prevent those from occurring?

Hard to say how to avoid these in general. Some of the covariance matrices that appear during the course of BO, especially in later iterations, can be quite ill conditioned (and potentially numerically not p.d.). These ill conditioned covariances are likely also the cause of the negative variances.

> In the quote above, are you suggesting that 5 objectives is too much for Botorch to handle? I am asking cuz I've read in the literature that MOBO can generally handle upto 15-ish objectives.

What literature suggests that? With that many outputs hypervolume approaches are typically not easily feasible, and pretty significant approximations need to be made or scalarization-based methods need to be used. Do you maybe mean that MOBO can handle 15 *input* (rather than outcome) dimensions? That has been a ballpark number for Bayesian Optimization for a while (though it's not true anymore with more modern methods that scale better). 

> Also, I don't believe my problem has a constraint, other than a simple box bound imposed on the input parameters that has been explicitly written out in the optimize_acqf_and_get_observation(model, train_x) method in my code :

What I was suggesting is that considering some of the outcomes as things you may constrain on, rather than adding them to the list of objectives to explore the Pareto Frontier for. This will increase scalability of the hypervolume computations.

**Reply by saitcakmak** [reply]
> In addition, in the output you can see two types of numerical warnings: a "not p.d., added jitter" warning and a "Negative variance values" warning. Any idea how to prevent those from occurring?

One thing that may help with these is to standardize the objective values before feeding them into the model. Some of your observed values are very small, which will lead to very small values after the mean & covariance matrix get unstandardized (they get standardized and unstandardized during training & inference under the hood since you're using `Standardize` outcome transform`).  The posterior object has its MVN in the raw objective space, which is where the "negative variance" warnings most likely come from.

**Reply by renjieli08** [reply]
Thank you both for answering. 

> With that many outputs hypervolume approaches are typically not easily feasible, and pretty significant approximations need to be made or scalarization-based methods need to be used.

Would you say that reducing the number of objectives to 3 would substantially improve the scalability of the hypervolume computations? I am trying to reformulate the problem to have only 3 objectives. 

>  One thing that may help with these is to standardize the objective values before feeding them into the model.

I think I have already standardized the objectives in my code. Is the following lines of code correct: 
``
SingleTaskGP(train_X, train_Y[..., i : i + 1], covar_module= covariance,
                      outcome_transform=Standardize(m=1), input_transform=Normalize(d=d))
``

Finally, I noticed that my posterior mean and 2 standard deviation confidence are almost identical, as seen in the output. I don't think this is a coincidence, and if not, what could have been wrong here? For instance, see:

Posterior mean and confidence region upper bound:
[[0.1053008990645664, 0.9554760433997954, 0.23714447045978354, 0.15496200862309312, 0.8384869527564519]] [[0.10556030090144049, 0.9554960433997954, 0.23718471349038067, 0.1551281692321272, 0.8385069527564519]]

**Reply by Balandat** [reply]
> Would you say that reducing the number of objectives to 3 would substantially improve the scalability of the hypervolume computations? I am trying to reformulate the problem to have only 3 objectives.

Yes, 3 is definitely tractable, 5 will require a good deal of approximations when using HV computations. 

> Is the following lines of code correct:

Yes that should do it

> Finally, I noticed that my posterior mean and 2 standard deviation confidence are almost identical, as seen in the output. I don't think this is a coincidence, and if not, what could have been wrong here?

I don't know if anything is wrong, this just suggests that your model have very low posterior uncertainty. This could explain the acquisition value being very small, since there may not be much uncertainty and and not much improvement to be had. I would recommend checking the model fits to make sure that those are well calibrated, see e.g. this tutorial: https://github.com/pytorch/botorch/blob/main/tutorials/batch_mode_cross_validation.ipynb

**Reply by renjieli08** [reply]
Hi Max,
Thanks for the explanation. 

After some iterations, the posterior mean and confidence are no longer close to each other in value. 

Other thing I'm experiencing is that after around 100 iterations, each iteration takes around 20 minutes to run and uses upto 60% CPU. I was able to determine that the function fit_gpytorch_mll(mll_ei) is consuming all the time and CPU resources. I understand that fit_gpytorch_mll(mll_ei) is fitting/updating the GP marginal likelihood. 

So is each call to fit_gpytorch_mll(mll_ei) taking 20 minutes to run a normal scenario? 

I am using a Windows workstation with 12 cores (24 logical processors), 128 GB of memory, the OS is Windows Server 2019 Datacenter. **BTW, the CUDA driver is broken so I can't use GPU for the tensor processings.** 

Thanks in advance.

**Reply by Balandat** [reply]
So 100 iterations means around 100 data points? Or are you generating large batches of candidates? While GPs don't scale particularly well with the number of observations, with 100 datapoints 20 minutes for doing a MAP estimate seems way to high, this should usually be on the order of a few seconds. What CPU does the machine have? One thing that could be going on is that the HW acceleration of the CPU may not be properly working. E.g. if you have an Intel CPU this would be MKL. Botorch makes heavy use of that via pytorch, so if this isn't working as intended this can slow things down significantly (which might be what's going on for you).

**Reply by renjieli08** [reply]
Hi Max,
Yeah 100 iterations means 100 calls to fit_gpytorch_mll(mll_ei) and the optimize_acqf_and_get_observation(model, train_x) method (since i'm generating a single candidate each iteration, this means 100 datapoints as well). The logic can be seen in my code:
```
    for iteration in range(1, N_BATCH + 1):
        #do one iteration
        print('\nStarting step No.{}'.format(iteration))

        steps_done += 1

        # fit the models
        fit_gpytorch_mll(mll_ei)

        # optimize and get new observation
        new_x_ei, new_obj_ei, acq, score = optimize_acqf_and_get_observation(model_ei, train_x_ei)
        ...
```

I've fixed the CUDA driver and now I'm using GPU instead of CPU. **Each iteration now takes around 3.5 mins**. Below screenshot is the usage of the GPU at the 400th iteration: 
![Screen Shot 2023-06-12 at 09 25 46](https://github.com/pytorch/botorch/assets/76676601/2bc670c6-2cfe-4661-8ac1-38000a4f4acb)

Is 3.5 mins each iteration an acceptable amount of time? I understand that the GPUs on our cluster are quite old.
BTW, I've also reduced the number of objectives from 5 to 3 per our discussion above. 

Thanks.

**Reply by Balandat** [reply]
Hmm this still sounds rather slow. Can you share the updated version of the code that you're running so I can see if I get comparable timings?

**Reply by renjieli08** [reply]
Hi max,

I messed up lol. Each botorch iteration takes about 20 sec now as opposed to 3.5 mins as I forgot to account for the time spent on running my physics simulations. 

So back to business. I mainly have concerns/doubts about the following method, since I now have three objectives and three constraints (so in total six outcomes, which yields a q by 6 tensor where first 3 columns are obj and last 3 are constraints):

```
def optimize_acqf_and_get_observation(model, train_x):
    """Optimizes the acquisition function, and returns a new candidate and a noisy observation."""

    #bounds for the x input
    high = [1000.0, 200., 200., 200., 0.15, 0.15, 0.3, 100.]
    low = [-1000.0, -200., -200., -200., -0.15, -0.15, -0.3, -100.]
    bounds = torch.tensor([low, high], dtype=torch.float64, device=device)

    train_x = normalize(train_x, bounds)

    #acquisition function (qNEHVI for multi-objective)
    acq_func = qNEHVI(
            model=model,
            ref_point= torch.zeros(3, dtype=torch.float64, device=device),  #lower bound for three objectives
            X_baseline=train_x,
            objective = IdentityMCMultiOutputObjective(outcomes = [0,1,2]),               #first 3 columns are objectives 
            constraints=[lambda Z: Z[..., -3], lambda Z: Z[..., -2], lambda Z: Z[..., -1]],  #last three columns are constraints
            prune_baseline=True,
            alpha=0.0,
    )

    #optimize acq. Higher raw_samples is better but more expensive.
    candidate, acq_value = optimize_acqf(
        acq_function = acq_func, bounds=bounds, q=1, num_restarts=5, raw_samples=128,
    )

    # observe new values
    new_x = candidate.detach()
    obj, score = env.step(new_x.tolist()[0])    #the objective function is a physics simulator set up using openAI gym
    obj = torch.tensor(obj, device=device).unsqueeze(0)
    new_y = obj[:, 2:5]
    #constraints
    con1 = obj[:, 0].unsqueeze(1) - 0   # Q factor > 0
    con2 = obj[:, 1].unsqueeze(1) - 0.923664    #abs(self.lam_goal - lam) < 100 nm
    con3 = 1.98 - obj[:, 3].unsqueeze(1)     #power < 0.99, conservation of energy.
    new_con = -1*torch.cat([con1, con2, con3], 1)

    return new_x, new_y, new_con, acq_value, score
```

In particular, could you please check if I got the following two lines correct? I feel pretty uneasy about the use of lambda in there. 
```
objective = IdentityMCMultiOutputObjective(outcomes = [0,1,2]),         #first 3 columns are objectives  
constraints=[lambda Z: Z[..., -3], lambda Z: Z[..., -2], lambda Z: Z[..., -1]],  #last three columns are constraints
```

**Reply by Balandat** [reply]
> I messed up lol. Each botorch iteration takes about 20 sec now as opposed to 3.5 mins as I forgot to account for the time spent on running my physics simulations.

Great, this sounds a lot more reasonable. 

> In particular, could you please check if I got the following two lines correct? I feel pretty uneasy about the use of lambda in there.

That seems ok to me - your constraints are that the last three elements of `Z` must be *greater* than zero, right?

**Reply by renjieli08** [reply]
Hi Max,

> That seems ok to me - your constraints are that the last three elements of Z must be greater than zero, right?

Yes, that's right. Here is how the constraints is defined, **where con1, con2, and con3 all need to be > 0.** 
```
#constraints
con1 = obj[:, 0].unsqueeze(1) - 0            # Q factor > 0
con2 = obj[:, 1].unsqueeze(1) - 0.923664    #abs(self.lam_goal - lam) < 100 nm
con3 = 1.98 - obj[:, 3].unsqueeze(1)     #power < 0.99, conservation of energy.
new_con = -1*torch.cat([con1, con2, con3], 1)
```

Is it correct to concatenate con1, con2, con3 side by side like I did above? 

Btw, my model is defined as below, where you will see that train_outcome (or the "Z" variable) is a tensor whose first 3 columns are obj and last 3 are constraints:
```
def initialize_model(train_X, train_Y, train_con):
    #initialize model
    train_outcome = torch.cat([train_Y, train_con], dim=-1)     #first 3 columns are obj and last 3 are constraints
    models = []
    for i in range(train_outcome.shape[-1]):   #shape is qx6
        models.append(
            SingleTaskGP(train_X, train_outcome[..., i : i + 1], covar_module= covariance,
                      outcome_transform=Standardize(m=1), input_transform=Normalize(d=d)
            )
        )
    model = ModelListGP(*models)
    mll = SumMarginalLogLikelihood(model.likelihood, model)
```

**Reply by Balandat** [reply]
Wait, when you define `con1 = obj[:, 0].unsqueeze(1) - 0`, is `obj[:, 0]` actually your first objective? Or is `obj` here some other outcome (that happens to sound like it contains your optimization objectives)? If this is indeed your objective and you just want to explore the outcome region where it's positive then you should encode this as the *reference point* in the acquisition function, rather than model it as a separate constraint.

**Reply by renjieli08** [reply]
The obj variable is the output from my physics simulator containing 5 quantities and I transform it into an 1 by 5 tensor as shown below: 
```
    # observe new values
    new_x = candidate.detach()
    obj, score = env.step(new_x.tolist()[0])    #the objective function is a physics simulator set up using openAI gym
    obj = torch.tensor(obj, device=device).unsqueeze(0)  # a 1 by 5 tensor of output from simulator 
    new_y = obj[:, 2:5].   #this my actual objective, which is the third, fourth and fifth element of the obj tensor
    #constraints
    con1 = obj[:, 0].unsqueeze(1) - 0   # con1 is the 1st element of the obj tensor
    con2 = obj[:, 1].unsqueeze(1) - 0.923664    #con2 is the 2nd element of the obj tensor
    con3 = 1.98 - obj[:, 3].unsqueeze(1)     #con3 is the 4th element of the obj tensor
    new_con = -1*torch.cat([con1, con2, con3], 1)
```
So basically, the obj tensor is NOT my actual objective, but rather a raw tensor containing a mixture of objectives and constraints which I split up according to the code above. Is there something off with the way I created my objectives and constraints?

> If this is indeed your objective and you just want to explore the outcome region where it's positive then you should encode this as the reference point in the acquisition function, rather than model it as a separate constraint.

I thought the ref_point argument only includes objectives and not constraints?

**Reply by Balandat** [reply]
So you're optimizing (as one of 3 objectives) the fourth element of `obj`, but you're also imposing a constraint on it (to be less than 1.98). Instead of doing that, you should encode the fact that you want it to be less than 1.98 in the respective dimension of the reference point. 

> I thought the ref_point argument only includes objectives and not constraints

That is right, but it looks like you're also constraining one of your objectives. And that should be done via the reference point. Otherwise you'd be building two GP models for the same outcome, and the algorithm probably also wouldn't explore the pareto frontier near 1.98 as well since with a constraint it'd be discouraged to explore there.

**Reply by renjieli08** [reply]
According to the tutorial at https://botorch.org/tutorials/constrained_multi_objective_bo, the reference point is defined as: 

"qNEHVI requires specifying a reference point, which is the lower bound on the objectives used for computing hypervolume. In this tutorial, we assume the reference point is known. In practice the reference point can be set 1) using domain knowledge to be slightly worse than the lower bound of objective values, where the lower bound is the minimum acceptable value of interest for each objective, or 2) using a dynamic reference point selection strategy."

Since my constraint is to be less than 1.98, which is sort of an upper bound, how should I transform the problem so it becomes a lower bound to be used in the reference point?

**Reply by Balandat** [reply]
Interesting - what's the reason you're providing this as an upper bound? This seems odd if you're maximizing this objective. Is this encoding some domain knowledge that values >= 1.98 are bad in some way (e.g. b/c they negatively affect some other outcome that you're not tracking explicitly)?

**Reply by renjieli08** [reply]
Yeah this a domain knowledge where values >= 1.98 are physically impossible and thus should always be avoided. I am basically maximizing this objective such that it's as large as possible while being less than 1.98.

**Reply by Balandat** [reply]
Hmm I see. This is a somewhat atypical situation so it's not supported super well out of the box, but you should be able to apply the outcome constraint as you initially suggested. Let me know how that goes.
